docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+
+Attaching `bs.string` to the polymorphic variant type makes its constructor
+compile to a string of the same name. This technique can be combined with the
+`bs.as` attributes to modify the strings produced from the polymorphic variant
+values, we will see an example of this just below.
+
+Aside from producing string values, Melange also offers `bs.int` to produce
+integer values:
+
+```ocaml
+external test_int_type :
+  ([ `on_closed | `on_open [@bs.as 20] | `in_bin ][@bs.int]) -> int
+  = "testIntType"
+  [@@bs.val]
+
+let value = test_int_type `on_open
+```
+
+In this example, `on_closed` will be encoded as 0, `on_open` will be 20 due to
+the attribute `bs.as` and `in_bin` will be 21.
+
+This code generates:
+
+```js
+var value = testIntType(20);
+```
+
+#### Using polymorphic variants to bind to event listeners
+
+Polymorphic variants can also be used to wrap event listeners, or any other kind
+of callback, for example:
+
+```ocaml
+type readline
+
+external on :
+  readline ->
+  ([ `close of unit -> unit | `line of string -> unit ][@bs.string]) ->
+  readline = "on"
+  [@@bs.send]
+
+let register rl =
+  rl |. on (`close (fun event -> ())) |. on (`line (fun line -> Js.log line))
+```
+
+This generates:
+
+```js
+function register(rl) {
+  return rl
+    .on("close", function($$event) {})
+    .on("line", function(line) {
+      console.log(line);
+    });
+}
+```
+
+#### Constant values as arguments
+
+Sometimes we want to call a JavaScript function and make sure one of the
+arguments is always constant. For this, the `[@bs.as]` attribute can be combined
+with the wildcard pattern `_`:
+
+```ocaml
+external process_on_exit : (_[@bs.as "exit"]) -> (int -> unit) -> unit
+  = "process.on"
+  [@@bs.val]
+
+let () =
+  process_on_exit (fun exit_code ->
+    Js.log ("error code: " ^ string_of_int exit_code))
+```
+
+This generates:
+
+```js
+process.on("exit", function (exitCode) {
+  console.log("error code: " + exitCode.toString());
+});
+```
+
+The `bs.as "exit"` and the wildcard `_` pattern together will tell Melange to
+compile the first argument of the JavaScript function to the string `"exit"`.
+
+You can also use any JSON literal by passing a quoted string to `bs.as`: `bs.as
+{json|true|json}` or `bs.as {json|{"name": "John"}|json}`.
+
+#### Binding to callbacks
+
+In OCaml, all functions have arity 1. This means that if you define a function
+like this:
+
+```ocaml
+let add x y = x + y
+```
+
+Its type will be `int -> int -> int`. This means that one can partially apply
+`add` by calling `add 1`, which will return another function expecting the
+second argument of the addition. This kind of functions are called "curried"
+functions, more information about currying in OCaml can be found in [this
+chapter](https://cs3110.github.io/textbook/chapters/hop/currying.html) of the
+"OCaml Programming: Correct + Efficient + Beautiful" book.
+
+Currying is problematic when binding to JavaScript, let’s say we have a function
```

Comment:
I don't like the word "problematic". We should probably just say "incompatible calling conventions", no?

Reply:
Removed "problematic" and expanded a bit in https://github.com/melange-re/documentation-site/pull/12/commits/127995137166ab4cc9c597b6ff9266fa2c7865e2.